lib/rust/ensogl/component/text/src/buffer.rs

```diff
@@ -18,129 +32,1143 @@ pub mod traits {
+        RangeLike::BufferRangeLocationColumn(*range)
+    }
+}
+
+impl From<&std::ops::Range<Byte>> for RangeLike {
+    fn from(range: &std::ops::Range<Byte>) -> Self {
+        RangeLike::RangeBytes(range.clone())
+    }
+}
+
+impl From<&RangeFull> for RangeLike {
+    fn from(range: &RangeFull) -> Self {
+        RangeLike::RangeFull(*range)
+    }
 }
 
-/// Generic setter for default value for metadata like colors, font weight, etc.
-trait DefaultSetter<T> {
-    /// Replace the default value of the metadata. The exact meaning of this function depends on the
-    /// provided data type. See implementations provided in the `style` module.
-    fn set_default(&self, data: T);
+
+
+// ====================
+// === LocationLike ===
+// ====================
+
+/// A location-like description. Any of the enum variants can be used to describe a location in
+/// text. There are conversions defined, so you never need to use this type explicitly.
+#[allow(missing_docs)]
+#[derive(Debug, Copy, Clone, From)]
+pub enum LocationLike {
+    LocationColumnLine(Location<Column, Line>),
+    LocationUBytesLine(Location<Byte, Line>),
+    LocationColumnViewLine(Location<Column, ViewLine>),
+    LocationUBytesViewLine(Location<Byte, ViewLine>),
+}
+
+impl Default for LocationLike {
+    fn default() -> Self {
+        LocationLike::LocationColumnLine(default())
+    }
 }
 
-impl Setter<Text> for Buffer {
-    fn replace(&self, range: impl enso_text::RangeBounds, text: Text) {
-        let range = self.crop_byte_range(range);
-        let size = text.byte_size();
-        self.text.replace(range, text);
-        self.style.set_resize_with_default(range, size);
+impl LocationLike {
+    /// Expand the [`LocationLike`] structure to a known location.
+    pub fn expand(self, buffer: &BufferModel) -> Location<Column, Line> {
+        match self {
+            LocationLike::LocationColumnLine(loc) => loc,
+            LocationLike::LocationUBytesLine(loc) => Location::from_in_context_snapped(buffer, loc),
+            LocationLike::LocationColumnViewLine(loc) =>
+                Location::from_in_context_snapped(buffer, loc),
+            LocationLike::LocationUBytesViewLine(loc) =>
+                Location::from_in_context_snapped(buffer, loc),
+        }
     }
 }
 
-impl Setter<&Text> for Buffer {
-    fn replace(&self, range: impl enso_text::RangeBounds, text: &Text) {
-        self.replace(range, text.clone())
+
+// ===================
+// === Conversions ===
+// ===================
+
+/// Perform conversion between two values. It is just like the [`From`] trait, but it performs the
+/// conversion in a "context". The "context" is an object containing additional information required
+/// to perform the conversion. For example, the context can be a text buffer, which is needed to
+/// convert byte offset to line-column location.
+#[allow(missing_docs)]
+pub trait FromInContext<Ctx, T> {
+    fn from_in_context(context: Ctx, arg: T) -> Self;
+}
```

Comment:
It is, but you can't reuse it because of Rust limitations for trait implementation between crates, unfortunately. I was fighting with it and finished with double definition :(